### PR TITLE
#167190540 Implement Follow And Unfollow User Feature

### PR DESCRIPTION
**Pull request by henryade**
#### What does this PR do?
It allows users to follow and unfollow other users, get all followers and get all users they follow.

#### Description of Task proposed in this pull request?
- create user follower model
- create follow user route
- create unfollow user route
- create get followers route
- create get followings route
- write end to end test on all route
- write model unit tests

#### How should this be manually tested (Quality Assurance)?
- clone repo and checkout to this branch
- signup two users and copy the first users token and the second user's username
### To Follow
- run `POST /api/v1/profiles/{username}/follow`add the first users token to the authorization header
### To get All followers
- run `GET /api/v1/profiles/followers` add the first users token to the authorization header
### To get All followings
- run `GET /api/v1/profiles/followings` add the first users token to the authorization header
### To Unfollow
- run `DELETE /api/v1/profiles/{username}/follow`add the first users token to the authorization header

#### What are the relevant pivotal tracker stories?
- Story ID:  [#167190540](https://www.pivotaltracker.com/story/show/167190540)

#### Any background context you want to add (Operations Impact)?
N/A

#### What I have learned working on this feature: 
- I learned how to write model tests and add custom error messages to models.
- I learned how to use sequelize magic methods. 

#### Screenshots: 
User Successfully follows another user:
Screenshot 2019-08-08 at 4 27 49 PM

User Successfully unfollow another user:
Screenshot 2019-08-08 at 4 31 34 PM

Other User View All Followers:
Screenshot 2019-08-08 at 4 31 03 PM

User View All Followings:
Screenshot 2019-08-08 at 4 28 22 PM

### CODE REVIEW

File: server/controllers/Followers.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+   * @returns {JSON} JSON object with details of new follower
+   */
+  static async follow(req, res) {
+    try {
+      const {
+        user: { id },
+        params: { username }
+      } = req;
+      const user = await User.findByUsername(username);
+      if (!user) {
```
Comment (dinobi) [inline server/controllers/Followers.js]: @henryade don't you think we should move these checks from lines 27-34 to another method inside this controller and use it here and also in the unfollow route handler, cos it looks like we are repeating code. In the future we could move to a helper, if we discover that someone else is trying to do this same thing.

File: server/controllers/Followers.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+   * @param {Object} req express request object
+   * @param {Object} res express response object
+   * @returns {JSON} JSON object with details of an unfollowed user
+   */
+  static async allFollowers(req, res) {
+    try {
+      const {
+        user: { id }
+      } = req;
+      const user = await User.findByPk(id);
```
Comment (dinobi) [inline server/controllers/Followers.js]: I thought we defined a `findByID`?

File: server/controllers/Followers.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      const followers = await user.getAllFollowers({
+        include: [
+          {
+            model: User,
+            as: 'follower'
+          }
+        ]
+      });
+      const users = followers.map((x) => {
```
Comment (dinobi) [inline server/controllers/Followers.js]: never use single characters for parameter names unless it's a counter, change to `x` to `follower`

File: server/controllers/Followers.js (hunk 1 above)
Comment (henryade) [reply server/controllers/Followers.js]: Okay. I will do this now.

File: server/controllers/Followers.js (hunk 2 above)
Comment (henryade) [reply server/controllers/Followers.js]: Yes. I will change this.

File: server/controllers/Followers.js (hunk 3 above)
Comment (henryade) [reply server/controllers/Followers.js]: okay. Noted.

**Review by Adekoreday — DISMISSED** [review]

**Review by Adekoreday — DISMISSED** [review]

**Review by abdulfatai360 — DISMISSED** [review]

**Review by JuwonAbiola — DISMISSED** [review]

File: server/database/migrations/20190805160738-create-user-follower.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+module.exports = {
```
Comment (chialuka) [inline server/database/migrations/20190805160738-create-user-follower.js]: To be consistent, could you please use ES6 syntax here and in the other migration files?

**Review by chialuka — DISMISSED** [review]

File: server/database/migrations/20190805160738-create-user-follower.js (hunk 4 above)
Comment (henryade) [reply server/database/migrations/20190805160738-create-user-follower.js]: Thank you

**Review by JuwonAbiola — DISMISSED** [review]

**Review by abdulfatai360 — DISMISSED** [review]

**Review by Adekoreday — DISMISSED** [review]

**Review by Adekoreday — DISMISSED** [review]

**Review by JuwonAbiola — DISMISSED** [review]

**Review by chialuka — DISMISSED** [review]
Good job.

**Review by dinobi — APPROVED** [review]

**Review by abdulfatai360 — APPROVED** [review]

**Review by Adekoreday — APPROVED** [review]

**Pull request merged**